platforms/nuttx/src/px4/stm/stm32_common/hrt/hrt.c

```diff
@@ -72,6 +72,23 @@
 #  define hrtinfo(x...)
 #endif
 
+#if !defined(CONFIG_BUILD_FLAT)
+#include <px4_platform_common/defines.h>
+#include <px4_platform/board_ctrl.h>
+#include <px4_platform_common/sem.h>
+
+static px4_sem_t g_wait_sem;
+static struct hrt_call *next_hrt_entry;
+
+void hrt_usr_call(void *arg)
+{
+	// This is called from hrt interrupt
+	next_hrt_entry = (struct hrt_call *)arg;
```

Comment:
This could drop events in case the thread can't keep up with the rate of triggering (or worse: the same entry could be called multiple times as the semaphore counter is increased). For the same reason `next_hrt_entry` needs to be atomic.

Reply:
This is something that I was thinking as well. I am just wondering if there is anything you can do to prevent it. The thread is set to highest priority, which means that nothing else than the interrupts can pre-empt it.

Eventually it is the same thing withe the flat-build irq handling; if the interrupts occur at higher rate than what can be handled, they will be dropped...

I will look at it again, thanks for the review!

Reply:
the next_hrt_entry is written in irq, and it cannot be pre-empted. The next_hrt_entry is only being read in hrt_ioctl, where it is theoretically possible that the pointer assgnment is not happening atomically and there is another interrupt in between. Good catch...  The struct content itself needs no protection, since the struct only allocated and managed in the client side, only the pointer is being passed around. So, I will just wrap the assignment in critical_section:

		flags = px4_enter_critical_section();
		*(struct hrt_call **)arg = next_hrt_entry;
		px4_leave_critical_section(flags);

Reply:
Exactly. Although in flat builds it would not drop, but rather just delay execution (and at some point drop things due to CPU overload).
To completely avoid this you would need to use a linked list/queue. I would not go that far but I see a real risk of this happening with only 1 entry. I'd add a fixed-size buffer of ~3 entries.
If you set the entry to null after reading it, you can also detect a drop from the worker thread (if woken up w/o valid entry) and print a message (once). If you find one entry is enough we can also leave it at that and should it ever become a problem we'll notice.

Reply:
> If you set the entry to null after reading it, you can also detect a drop from the worker thread (if woken up w/o valid entry) and print a message (once). If you find one entry is enough we can also leave it at that and should it ever become a problem we'll notice.

This is good input; I will modify this part according to your suggestions, thanks!

Reply:
@bkueng I added a 3-level stack for the user space hrt calls, and the error message as you suggested